library/common.h

```diff
@@ -68,6 +69,42 @@ extern void (*mbedtls_test_hook_test_fail)( const char * test, int line, const c
  */
 #define MBEDTLS_ALLOW_PRIVATE_ACCESS
 
+/** Return an offset into a buffer.
+ *
+ * This is just the addition of an offset to a pointer, except that this
+ * function also accepts an offset of 0 into a buffer whose pointer is null.
+ *
+ * \param p     Pointer to a buffer of at least n bytes.
+ *              This may be \p NULL if \p n is zero.
+ * \param n     An offset in bytes.
+ * \return      Pointer to offset \p n in the buffer \p p.
+ *              Note that this is only a valid pointer if the size of the
+ *              buffer is at least \p n + 1.
+ */
+static inline unsigned char *mbedtls_buffer_offset(
```

Comment:
Wouldn't this be better as a macro, so that the two definitions can be combined, and it can accept pointers to any type?

Reply:
> it can accept pointers to any type

I made it a function partly because I don't want it to accept pointers of any type. This is only intended for data buffers, where the pointer can be null for an empty buffer.

What would be the benefit of accepting arbitrary types?

Reply:
The benefit is minor I guess if there are no current use-cases for other types, but equally there's no particular reason not to allow it. It does improve readability by not duplicating the function definition for const/non-const.

Reply:
> there's no particular reason not to allow it

There is: this makes it possible for parts of the code to accept a null pointer when a null pointer was not intended to be valid (and will cause problems later). So it should be used sparingly. I don't want to encourage making things nullable all over the place: [nullability is dangerous](https://en.wikipedia.org/wiki/Null_pointer#History).

Arguments for making it a macro:

* Saves duplication. But only two copies: there won't be more.
* Saves having to select the correct name when using it.

Arguments for making it a function:

* Prevents using it on unintended types.
* Behaves transparently if the arguments have side effects.

The pro-function arguments are much stronger than the pro-macro arguments so I don't intend to change to a macro unless there are more pro-macro argumens.